fold/fold.go

```diff
@@ -0,0 +1,895 @@
+		}
+	}
+	return string(result)
+}
+
+// Find and return the lowest free energy structure in the subsequence starting
+// at start and terminating at end.
+//
+// Figure 2B in Zuker and Stiegler, 1981
+// Args:
+//
+//		seq: The sequence being folded
+//		start: The start index
+//		end: The end index (inclusive)
+//	 foldContext: The FoldContext for this sequence
+//
+// Returns the free energy for the subsequence from start to end
+func W(start, end int, foldContext FoldContext) (NucleicAcidStructure, error) {
+	if !foldContext.W[start][end].Equal(DefaultStructure) {
+		return foldContext.W[start][end], nil
+	}
+
+	if end-start < 4 {
+		foldContext.W[start][end] = InvalidStructure
+		return foldContext.W[start][end], nil
+	}
+
+	w1, err := W(start+1, end, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+	w2, err := W(start, end-1, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+	w3, err := V(start, end, foldContext)
+	if err != nil {
+		return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+	}
+
+	w4 := InvalidStructure
+	for k := start + 1; k < end-1; k++ {
+		w4_test, err := Multibranch(start, k, end, foldContext, false)
+		if err != nil {
+			return DefaultStructure, fmt.Errorf("w: subsequence (%d, %d): %w", start, end, err)
+		}
+
+		if w4_test.Valid() && w4_test.Energy < w4.Energy {
+			w4 = w4_test
+		}
+	}
+
+	wret := minStruct(w1, w2, w3, w4)
+	foldContext.W[start][end] = wret
+	return wret, nil
+}
+
+// Find, store and return the minimum free energy of the structure between start
+// and end.
+//
+// If start and end don't bp, store and return INF.
+// See: Figure 2B of Zuker, 1981
+// Args:
+//
+//		start: The start index
+//		end: The end index (inclusive)
+//	 foldContext: The FoldingContext for this sequence
+//
+// Returns the minimum energy folding structure possible between start and end on seq
+func V(start, end int, foldContext FoldContext) (NucleicAcidStructure, error) {
+	if !foldContext.V[start][end].Equal(DefaultStructure) {
+		return foldContext.V[start][end], nil
+	}
+
+	// the ends must basepair for V(start,end)
+	if foldContext.Energies.Complement[foldContext.Seq[start]] != foldContext.Seq[end] {
+		foldContext.V[start][end] = InvalidStructure
+		return foldContext.V[start][end], nil
+	}
+	// if the basepair is isolated, and the seq large, penalize at 1,600 kcal/mol
+	// heuristic for speeding this up
+	// from https://www.ncbi.nlm.nih.gov/pubmed/10329189
+	isolated_outer := true
+	if start > 0 && end < len(foldContext.Seq)-1 {
+		isolated_outer = foldContext.Energies.Complement[foldContext.Seq[start-1]] != foldContext.Seq[end+1]
+	}
+	isolated_inner := foldContext.Energies.Complement[foldContext.Seq[start+1]] != foldContext.Seq[end-1]
+
+	if isolated_outer && isolated_inner {
+		foldContext.V[start][end] = NucleicAcidStructure{Energy: 1600}
```

Comment:
why this magic nummber of `1600` for energy?

Reply:
Apparently it's a heuristic from https://www.ncbi.nlm.nih.gov/pubmed/10329189

Moving to a const.